Supports for Riscv

Does `buck2` currently have supports for `Riscv` and to what extent? If not, how do I add support? Some guidance would be appreciated, thanks.

I tried, but there are some error.
```
nju@njuyxz:~/riscv-demo$ file $(buck2 build //:hello_riscv --show-full-output | grep hello_word | awk '{print $NF}')
Starting new buck2 daemon...
Connected to new buck2 daemon.
Action failed: root//:hello_riscv (rustc link [pic])
Local command returned non-zero exit code 1
Reproduce locally: `env -- 'BUCK_SCRATCH_PATH=buck-out/v2/tmp/root/6fc2dc79ee3d66c2/rustc/_buck_c5a998e79d3a718a' /usr/b ...<omitted>... -out/v2/gen/root/904931f735703749/__hello_riscv__/bin-pic-static_pic-link/hello_riscv-link-diag.args (run `buck2 log what-failed` to get the full command)`
stdout:
stderr:
error: linking with `buck-out/v2/gen/root/904931f735703749/__hello_riscv__/linker_wrapper.sh` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/bin:/home/nju/.cursor-server/bin/906121b8c0bdf041c14a15dac228e66ab5505260/bin/remote-cli:/home/nju/.local/bin:/home/nju/.cargo/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/usr/lib/wsl/lib:/mnt/c/Users/yxz/AppData/Local/Programs/cursor/resources/app/bin:/mnt/c/WINDOWS/system32:/mnt/c/WINDOWS:/mnt/c/WINDOWS/System32/Wbem:/mnt/c/WINDOWS/System32/WindowsPowerShell/v1.0/:/mnt/c/WINDOWS/System32/OpenSSH/:/mnt/c/Program Files/NVIDIA Corporation/NVIDIA app/NvDLISR:/mnt/c/Program Files (x86)/NVIDIA Corporation/PhysX/Common:/mnt/c/Users/yxz/AppData/Local/Programs/Python/Python311/Scripts/:/mnt/c/Users/yxz/AppData/Local/Programs/Python/Python311/:/mnt/c/Users/yxz/AppData/Local/Microsoft/WindowsApps:/mnt/c/Users/yxz/AppData/Local/GitHubDesktop/bin:/mnt/d/Software/驱动/DriverGenius:/mnt/d/Software/驱动/DriverGenius/ksoft:/mnt/c/Users/yxz/AppData/Local/Programs/cursor/resources/app/bin:/mnt/c/Program Files (x86)/Windows Kits/10/Windows Performance Toolkit/:/mnt/c/Users/yxz/AppData/Local/Programs/Python/Python311/Scripts/:/mnt/c/Users/yxz/AppData/Local/Programs/Python/Python311/:/mnt/c/Users/yxz/AppData/Local/Microsoft/WindowsApps:/mnt/d/Software/Code/Code/Microsoft VS Code/bin:/mnt/c/Users/yxz/AppData/Local/GitHubDesktop/bin:/snap/bin:/usr/local/llvm/bin:/home/nju/.cursor-server/data/User/globalStorage/github.copilot-chat/debugCommand:/usr/local/llvm/bin" VSLANG="1033" "buck-out/v2/gen/root/904931f735703749/__hello_riscv__/linker_wrapper.sh" "/home/nju/riscv-demo/buck-out/v2/tmp/root/6fc2dc79ee3d66c2/rustc/_buck_c5a998e79d3a718a/rustczJaaAy/symbols.o" "buck-out/v2/gen/root/904931f735703749/__hello_riscv__/bin-pic-static_pic-link/extras/hello_riscv/hello_riscv.hello_riscv.ffe23f9e54638fb5-cgu.0.rcgu.o" "buck-out/v2/gen/root/904931f735703749/__hello_riscv__/bin-pic-static_pic-link/extras/hello_riscv/hello_riscv.2mwofkz0oxl0z6ge.rcgu.o" "-Wl,--as-needed" "-L" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib" "-Wl,-Bstatic" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libstd-61a615288b660726.rlib" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libpanic_unwind-0763dd7f9631b484.rlib" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libobject-4632d96907a11c64.rlib" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libmemchr-f3974f00bc50325a.rlib" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libaddr2line-62051ffaf5460ac1.rlib" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libgimli-f93c3fed87fa51ae.rlib" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/librustc_demangle-48202346b7d7eddb.rlib" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libstd_detect-e40704e9114f948e.rlib" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libhashbrown-cd9ce99a6cc4cc77.rlib" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/librustc_std_workspace_alloc-42584d16403e4647.rlib" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libminiz_oxide-5af68ebaee708312.rlib" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libadler-47bcb2b3ed8fa6f4.rlib" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libunwind-3f6f1dff5becaf2a.rlib" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libcfg_if-975290e21194ce06.rlib" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/liblibc-ec7b5fcbc22b0eba.rlib" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/liballoc-ed790f12ae67e4bf.rlib" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/librustc_std_workspace_core-4fc1f45b8fe2f426.rlib" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libcore-1b62315ba7406f5d.rlib" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib/libcompiler_builtins-9d4603e742ceb515.rlib" "-Wl,-Bdynamic" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/home/nju/.rustup/toolchains/nightly-2023-11-13-x86_64-unknown-linux-gnu/lib/rustlib/riscv64gc-unknown-linux-gnu/lib" "-o" "buck-out/v2/gen/root/904931f735703749/__hello_riscv__/hello_riscv" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-nodefaultlibs" "-target" "riscv64gc-unknown-linux-gnu" "@buck-out/v2/gen/root/904931f735703749/__hello_riscv__/bin-pic-static_pic-link/__hello_riscv-link_linker_args.txt"
  = note: clang: error: unknown target triple 'riscv64gc-unknown-linux-gnu', please use -triple or -arch
          


error: aborting due to previous error


Build ID: 16d9644c-c859-4ebf-a706-f95570907443
Loading targets.   Remaining     0/13                                                                                                                          110 targets declared
Analyzing targets. Remaining     0/49                                                                                                                          97 actions, 133 artifacts declared
Executing actions. Remaining     0/7                                                                                                                           0.1s exec time total
Command: build.    Finished 1 local                                                                                                                                                                                                 
Time elapsed: 0.3s
BUILD FAILED
Failed to build 'root//:hello_riscv (prelude//platforms:default#904931f735703749)'
Usage: file [-bcCdEhikLlNnprsSvzZ0] [--apple] [--extension] [--mime-encoding]
            [--mime-type] [-e <testname>] [-F <separator>]  [-f <namefile>]
            [-m <magicfiles>] [-P <parameter=value>] <file> ...
       file -C [-m <magicfiles>]
       file [--help]
```
It looks like it compiles successfully, but the link fails. I have installed the target:
```
rustup target add riscv64gc-unknown-linux-gnu
```
I don't have experience with cross-platform linking, so I'm not quite sure why. Is additional configuration required?

Here is the demo project I attempted to compile: https://github.com/xizheyin/buck2-demo

--- a/toolchains/BUCK
+++ b/toolchains/BUCK
@@ -1,17 +1,23 @@
-load(":riscv_rust_toolchain.bzl", "riscv_rust_toolchain")
+load(":riscv_rust_toolchain.bzl", "riscv_cxx_toolchain", "riscv_rust_toolchain")
 load("@prelude//toolchains:cxx.bzl", "system_cxx_toolchain")
 load("@prelude//toolchains:python.bzl", "system_python_bootstrap_toolchain")
 
-system_cxx_toolchain(
+riscv_cxx_toolchain(
     name = "cxx",
+    base = ":cxx_host",
+    linker = "riscv64-linux-gnu-gcc",
+    linker_type = "gnu",
     visibility = ["PUBLIC"],
+)
+
+system_cxx_toolchain(
+    name = "cxx_host",
 )
 
 system_python_bootstrap_toolchain(
     name = "python_bootstrap",
     visibility = ["PUBLIC"],
 )
-
 
 riscv_rust_toolchain(
     name = "rust",

--- a/toolchains/riscv_rust_toolchain.bzl
+++ b/toolchains/riscv_rust_toolchain.bzl
@@ -1,7 +1,41 @@
+load("@prelude//cxx:cxx_toolchain_types.bzl", "CxxPlatformInfo", "CxxToolchainInfo", "LinkerInfo", "LinkerType")
 load("@prelude//rust:rust_toolchain.bzl", "PanicRuntime", "RustToolchainInfo")
 
+def _riscv_cxx_toolchain_impl(ctx: AnalysisContext) -> list[Provider]:
+    base = ctx.attrs.base[CxxToolchainInfo]
 
+    linker_info = {
+        field: getattr(base.linker_info, field)
+        for field in dir(base.linker_info)
+    } | dict(
+        linker = RunInfo(ctx.attrs.linker),
+        linker_flags = ctx.attrs.linker_flags,
+        type = LinkerType(ctx.attrs.linker_type),
+    )
 
+    cxx_toolchain_info = {
+        field: getattr(base, field)
+        for field in dir(base)
+    } | dict(
+        linker_info = LinkerInfo(**linker_info),
+    )
+
+    return [
+        DefaultInfo(),
+        ctx.attrs.base[CxxPlatformInfo],
+        CxxToolchainInfo(**cxx_toolchain_info),
+    ]
+
+riscv_cxx_toolchain = rule(
+    impl = _riscv_cxx_toolchain_impl,
+    attrs = {
+        "base": attrs.toolchain_dep(providers = [CxxPlatformInfo, CxxToolchainInfo]),
+        "linker": attrs.string(),
+        "linker_flags": attrs.list(attrs.arg(), default = []),
+        "linker_type": attrs.enum(LinkerType.values()),
+    },
+    is_toolchain_rule = True,
+)
 
 def _riscv_rust_toolchain_impl(ctx):
     return [

